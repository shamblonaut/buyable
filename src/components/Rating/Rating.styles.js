import styled from "styled-components";
import { LucideStar } from "lucide-react";

export const Container = styled.div`
  display: flex;
  align-items: center;
  padding: 8px 0;
`;

export const Stars = styled.div`
  display: grid;
  grid-template-columns: repeat(5, 1fr);
  gap: 4px;
`;

export const Star = styled.div`
  position: relative;
`;

export const StarShell = styled(LucideStar)`
  fill: #d8d8d8;
  stroke: none;
`;

export const FillMask = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  overflow: hidden;
  width: ${(props) => props.$percentage}%;
`;

export const StarFill = styled(LucideStar)`
  max-width: min-content;
  fill: yellow;
  stroke: 1px;
`;

export const Details = styled.span`
  display: flex;
  gap: 8px;
  margin-left: 8px;
`;

export const RatingValue = styled.p`
  font-weight: 500;
`;

export const RatingCount = styled.p`
  font-weight: 300;
`;
