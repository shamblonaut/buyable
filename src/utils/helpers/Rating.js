export const getRatingArray = (rating) => {
  if (rating > 5.0) {
    throw new Error("Invalid Rating");
  }

  let ratingArray = [];
  for (let i = 0; i < 5; i++) {
    const remainder = rating - i;
    ratingArray.push(
      remainder > 1 ? 1 : remainder > 0 ? Number(remainder.toFixed(1)) : 0,
    );
  }

  return ratingArray;
};
