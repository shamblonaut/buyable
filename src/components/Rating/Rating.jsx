import { getRatingArray } from "@/utils/helpers";

import {
  Container,
  Stars,
  Star,
  StarShell,
  FillMask,
  StarFill,
  Details,
  RatingValue,
  RatingCount,
} from "./Rating.styles";

const Rating = ({ rating, count }) => {
  return (
    <Container>
      <Stars>
        {getRatingArray(rating).map((starValue, index) => (
          <Star key={index}>
            <StarShell />
            <FillMask $percentage={starValue * 100}>
              <StarFill />
            </FillMask>
          </Star>
        ))}
      </Stars>
      <Details>
        <RatingValue>{rating}</RatingValue>
        <RatingCount>({count})</RatingCount>
      </Details>
    </Container>
  );
};

export default Rating;
