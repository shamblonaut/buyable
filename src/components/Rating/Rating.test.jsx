import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";

import Rating from "./Rating";

const mockRating = 3.2;
const mockCount = 256;

vi.mock("./Rating.styles", async (importOriginal) => {
  const actual = await importOriginal();

  const FillMask = actual.FillMask;
  return {
    ...actual,
    StarShell: () => <div data-testid="star-shell"></div>,
    FillMask: ({ children, ...props }) => (
      <FillMask data-testid="fill-mask" {...props}>
        {children}
      </FillMask>
    ),
  };
});

describe("Rating component", () => {
  it("renders correctly", () => {
    const { container } = render(
      <Rating rating={mockRating} count={mockCount} />,
    );

    expect(container).toMatchSnapshot();
  });

  it("has exactly 5 shells", () => {
    render(<Rating rating={mockRating} count={mockCount} />);

    expect(screen.getAllByTestId("star-shell").length).toBe(5);
  });

  it("sets the correct width for each star", () => {
    render(<Rating rating={mockRating} count={mockCount} />);

    const fillMaskList = screen.getAllByTestId("fill-mask");

    expect(fillMaskList.length).toBe(5);
    expect(fillMaskList[0]).toHaveStyle({ width: "100%" });
    expect(fillMaskList[1]).toHaveStyle({ width: "100%" });
    expect(fillMaskList[2]).toHaveStyle({ width: "100%" });
    expect(fillMaskList[3]).toHaveStyle({ width: "20%" });
    expect(fillMaskList[4]).toHaveStyle({ width: "0%" });
  });
});
