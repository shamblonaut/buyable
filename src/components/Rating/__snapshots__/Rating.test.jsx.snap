// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Rating component > renders correctly 1`] = `
<div>
  <div
    class="sc-bRKDuR fIviup"
  >
    <div
      class="sc-hvigdm ktpWWa"
    >
      <div
        class="sc-fhHczv PWdic"
      >
        <div
          data-testid="star-shell"
        />
        <div
          class="sc-dTvVRJ gofgAT"
          data-testid="fill-mask"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-star sc-jwTyAe bmcmzb"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
            />
          </svg>
        </div>
      </div>
      <div
        class="sc-fhHczv PWdic"
      >
        <div
          data-testid="star-shell"
        />
        <div
          class="sc-dTvVRJ gofgAT"
          data-testid="fill-mask"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-star sc-jwTyAe bmcmzb"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
            />
          </svg>
        </div>
      </div>
      <div
        class="sc-fhHczv PWdic"
      >
        <div
          data-testid="star-shell"
        />
        <div
          class="sc-dTvVRJ gofgAT"
          data-testid="fill-mask"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-star sc-jwTyAe bmcmzb"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
            />
          </svg>
        </div>
      </div>
      <div
        class="sc-fhHczv PWdic"
      >
        <div
          data-testid="star-shell"
        />
        <div
          class="sc-dTvVRJ epAoBQ"
          data-testid="fill-mask"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-star sc-jwTyAe bmcmzb"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
            />
          </svg>
        </div>
      </div>
      <div
        class="sc-fhHczv PWdic"
      >
        <div
          data-testid="star-shell"
        />
        <div
          class="sc-dTvVRJ fXtCnq"
          data-testid="fill-mask"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-star sc-jwTyAe bmcmzb"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
            />
          </svg>
        </div>
      </div>
    </div>
    <span
      class="sc-hjsuWn bUVtru"
    >
      <p
        class="sc-jJLAfE iFSvBN"
      >
        3.2
      </p>
      <p
        class="sc-hwkwBN iUfLLG"
      >
        (
        256
        )
      </p>
    </span>
  </div>
</div>
`;
